File: src/extensions/flags.ts
import { err } from '../types.js';
import { helperFailure, helperSuccess, type HelperResult } from './types.js';

export interface ParsedCountFlag {
  count: number;
  rest: string[];
}

export function parseCountFlag(
  commandName: string,
  args: string[],
  defaultCount: number,
): HelperResult<ParsedCountFlag> {
  if (args[0] !== '-n') {
    return helperSuccess({
      count: defaultCount,
      rest: [...args],
    });
  }

  if (args.length < 2) {
    return helperFailure(err(`${commandName}: missing value for -n`));
  }

  const count = Number.parseInt(args[1] ?? '', 10);
  if (!Number.isFinite(count)) {
    return helperFailure(err(`${commandName}: invalid integer for -n: ${args[1]}`));
  }

  return helperSuccess({
    count,
    rest: args.slice(2),
  });
}

File: python/src/onetool/extensions/flags.py
from __future__ import annotations

from dataclasses import dataclass

from ..types import err
from .types import HelperResult, helper_failure, helper_success


@dataclass(frozen=True, slots=True)
class ParsedCountFlag:
    count: int
    rest: list[str]


def parse_count_flag(
    command_name: str,
    args: list[str],
    default_count: int,
) -> HelperResult[ParsedCountFlag]:
    if not args or args[0] != "-n":
        return helper_success(ParsedCountFlag(count=default_count, rest=list(args)))

    if len(args) < 2:
        return helper_failure(err(f"{command_name}: missing value for -n"))

    try:
        count = int(args[1], 10)
    except ValueError:
        return helper_failure(err(f"{command_name}: invalid integer for -n: {args[1]}"))

    return helper_success(ParsedCountFlag(count=count, rest=args[2:]))
